src/t_list.c

```diff
@@ -125,17 +112,31 @@ void listTypeTryConvertQuicklist(robj *o, int shrinking, beforeConvertCB fn, voi
     /* Invoke callback before conversion. */
     if (fn) fn(data);
 
-    listTypeConvertToListpack(o);
+    /* Extract the listpack from the unique quicklist node,
+     * then reset it and release the quicklist. */
+    o->ptr = ql->head->entry;
+    ql->head->entry = NULL;
+    quicklistRelease(ql);
+    o->encoding = OBJ_ENCODING_LISTPACK;
 }
 
-void listTypeTryConversionForShrinking(robj *o, beforeConvertCB fn, void *data) {
-    if (o->encoding == OBJ_ENCODING_QUICKLIST) {
-        listTypeTryConvertQuicklist(o, 1, fn, data);
-    } else if (o->encoding == OBJ_ENCODING_LISTPACK) {
-        /* Nothing to do */
-    } else {
+/* Check if the list needs to be converted to appropriate encoding due to
+ * growing, shrinking or other cases.
+ *
+ * 'lct' can be one of the following values:
+ * LIST_CONV_GROWING   - Only consider converting from listpack to quicklist
+ * LIST_CONV_SHRINKING - Only consider converting from quicklist to listpack
```

Comment:
that's not really true (the "only consider" part), we do call both conversion functions regardless of this argument.
one uses it just instead of checking argv!=NULL, so it doesn't need that argument at all.
the other uses it to decide on the threshold (to avoid resonance).

i think we should remove the `growing` argument from the first one, and that we should document this enum something like:
* UNKNOWN - used after we built a new list, and we want to let the function decide on the best encoding for that list.
* GROWING - used before or right after adding elements to the list, in which case we are likely to only consider converting from listpack to quicklist.
* SHRINKING - used after removing an element from the list, in which case we wanna consider conversion from quicklist to listpack. When we know we're shrinking, we use a lower (more strict) threshold in order to avoid repeated conversions on every list change.

Reply:
Done.